src/common/type/pack.h

```diff
@@ -0,0 +1,505 @@
+cumulative field ID. At the end of a container the numbering will continue from wherever the outer container left off.
+***********************************************************************************************************************************/
+#ifndef COMMON_TYPE_PACK_H
+#define COMMON_TYPE_PACK_H
+
+/***********************************************************************************************************************************
+Minimum number of extra bytes to allocate for packs that are growing or are likely to grow
+***********************************************************************************************************************************/
+#ifndef PACK_EXTRA_MIN
+    #define PACK_EXTRA_MIN                                          128
+#endif
+
+/***********************************************************************************************************************************
+Object types
+***********************************************************************************************************************************/
+#define PACK_READ_TYPE                                              PackRead
+#define PACK_READ_PREFIX                                            pckRead
+
+typedef struct PackRead PackRead;
+
+#define PACK_WRITE_TYPE                                             PackWrite
+#define PACK_WRITE_PREFIX                                           pckWrite
+
+typedef struct PackWrite PackWrite;
+
+#include "common/io/read.h"
+#include "common/io/write.h"
+#include "common/type/string.h"
+
+/***********************************************************************************************************************************
+Pack data type
+***********************************************************************************************************************************/
+typedef enum
+{
+    pckTypeUnknown = 0,
+    pckTypeArray,
+    pckTypeBin,
+    pckTypeBool,
+    pckTypeI32,
+    pckTypeI64,
+    pckTypeObj,
+    pckTypePtr,
+    pckTypeStr,
+    pckTypeTime,
+    pckTypeU32,
+    pckTypeU64,
+} PackType;
+
+/***********************************************************************************************************************************
+Read Constructors
+***********************************************************************************************************************************/
+PackRead *pckReadNew(IoRead *read);
+PackRead *pckReadNewBuf(const Buffer *buffer);
+
+/***********************************************************************************************************************************
+Read Functions
+***********************************************************************************************************************************/
+typedef struct PackIdParam
+{
+    VAR_PARAM_HEADER;
+    unsigned int id;
+} PackIdParam;
+
+// Read next field
+bool pckReadNext(PackRead *this);
```

Comment:
pckReadNext() is intended for external use (see hrnPackToStr()) to do generic reads and also to read arrays (see tests).

I can certainly add to the comments here.

Reply:
If it has limitations, they should be called out in the interface (pack.h) comments so that it is clear to a user how/when to use/not to use this function.

Reply:
Agreed, will do once I can work on the branch again.

Reply:
Added in fa5bff95a5d5.